src/shared/concurrent-lru-cache.ts

```diff
@@ -152,33 +152,45 @@ export class ConcurrentLRUCacheWithContext<Context, Key extends ValidKey, Value>
             const value = result[key];
             if (value !== undefined) {
               this.cache.set(storableKey, { lastUpdated: Date.now(), value });
+              return value;
             }
+            return undefined;
           })
-          .catch(() => {})
+          .catch(() => undefined)
           .finally(() => this.beingCalculated.delete(storableKey));
         this.beingCalculated.set(storableKey, promise);
       }
     }
 
     // Wait for all calculations
-    const calculationPromises = notInCache.map((key) => this.beingCalculated.get(storableKeys[key]));
-    await Promise.all(calculationPromises);
+    const calculationPromises = notInCache.map<Promise<[Key, Value | undefined]>>(async (key) => [
+      key,
+      await this.beingCalculated.get(storableKeys[key]),
+    ]);
+    const calculated = Object.fromEntries(await Promise.all(calculationPromises)) as Record<Key, Value | undefined>;
 
     const nowAgain = Date.now();
     const useCachedValueIfCalculationFailed = ({ lastUpdated }: { lastUpdated: Timestamp }) =>
       options.useCachedValueIfCalculationFailed === 'always' || lastUpdated >= nowAgain - ms(options.useCachedValueIfCalculationFailed.ifUnder);
 
     // Check all values again
     for (const key of notInCache) {
-      const storableKey = storableKeys[key];
-      const valueInCache = this.cache.get(storableKey);
-      if (valueInCache) {
-        if (useCachedValueIfCalculationFailed(valueInCache)) {
-          // If can use it, then add it to result
-          result[key] = valueInCache.value;
-        } else {
-          // If not, then delete the value
-          this.cache.delete(storableKey);
+      // Check if we could calculate it
+      const calculatedValue = calculated[key];
+      if (calculatedValue !== undefined) {
+        result[key] = calculatedValue;
+      } else {
+        // If we couldn't calculate it, check if we had an old value stored in the cache we can still use
+        const storableKey = storableKeys[key];
+        const valueInCache = this.cache.get(storableKey);
```

Comment:
This cache has two different configs: 
* A value that determines when the config stops being valid
* Another value that determines that if we failed to calculate a value, we can still use the previous cached one

So for example, we can say values are valid for 5min, but, if the calculation failed, we can re-use the last cached value for up to 1h. So here, since we failed to calculate a value, we are checking the cache for the second config